### PR TITLE
Added gan process | along with example notebook

### PR DESCRIPTION
**Pull request by arunavo4**
@notadamking Refactored everything according to the new architecture

### CODE REVIEW

File: tensortrade/stochastic/processes/gan.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+        self._seq_length = kwargs.get('seq_length', 10)
+        self._hidden_units_g = kwargs.get('hidden_units_g', 150)
+        self._hidden_units_d = kwargs.get('hidden_units_d', 150)
+        self._num_generated_features = kwargs.get('num_generated_features', 1)
+        self._output_shape = kwargs.get('output_shape', (self._times_to_generate, 5, 1))
+        self._initialize_gan()
+        self._train_gan()
+        self.generate_price_history()
+
+    def _initialize_placeholders(self):
```
Comment (mwbrulhardt) [inline tensortrade/stochastic/processes/gan.py]: Since we are using tensorflow 2.0 we don't have to use placeholders anymore.

File: tensortrade/stochastic/processes/gan.py [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+        self._Z = tf.placeholder(tf.float32, [self._batch_size, self._seq_length, self._latent_dim])
+        self._W_out_G = tf.Variable(tf.truncated_normal([self._hidden_units_g, self._num_generated_features]))
+        self._b_out_G = tf.Variable(tf.truncated_normal([self._num_generated_features]))
+
+        self._X = tf.placeholder(tf.float32, [self._batch_size, self._seq_length, self._num_generated_features])
+        self._W_out_D = tf.Variable(tf.truncated_normal([self._hidden_units_d, 1]))
+        self._b_out_D = tf.Variable(tf.truncated_normal([1]))
+
+    @staticmethod
+    def sample_Z(batch_size, seq_length, latent_dim):
```
Comment (mwbrulhardt) [inline tensortrade/stochastic/processes/gan.py]: This doesn't need to be a method because it is only one line of code.

File: tensortrade/stochastic/processes/gan.py [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+
+        return dataset
+
+    @staticmethod
+    def get_batch(samples, batch_size, batch_idx):
+        start_pos = batch_idx * batch_size
+        end_pos = start_pos + batch_size
+        return samples[start_pos:end_pos]
+
+    def generator(self, z, c):
```
Comment (mwbrulhardt) [inline tensortrade/stochastic/processes/gan.py]: Can we have this more generalized so that if I have a custom generator and a custom discriminator I can just give those as input to this class and it can train them?

File: tensortrade/stochastic/processes/gan.py [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+            X_mb = self.get_batch(self._samples, self._batch_size, batch_idx + d + offset)
+            X_mb = X_mb.reshape(self._batch_size, self._seq_length, 1)
+            Y_mb = self.get_batch(self._samples, self._batch_size, batch_idx + d + offset)
+            _, D_loss_curr = self._sess.run([self._D_solver, self._D_loss],
+                                            feed_dict={self._CD: Y_mb, self._CG: Y_mb, self._X: X_mb,
+                                                       self._Z: self.sample_Z(self._batch_size, self._seq_length,
+                                                                              self._latent_dim)})
+        return D_loss_curr
+
+    def _initialize_gan(self):
```
Comment (mwbrulhardt) [inline tensortrade/stochastic/processes/gan.py]: This training needs to be simplified because it is written with tensorflow 1 and there has been much improvement in tensorflow 2.0 to simplify all of this.

File: tensortrade/stochastic/processes/gan.py [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+
+        generator_vars = [v for v in tf.trainable_variables() if v.name.startswith('generator')]
+        discriminator_vars = [v for v in tf.trainable_variables() if v.name.startswith('discriminator')]
+
+        D_loss_real = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(logits=d_logit_real,
+                                                                             labels=tf.ones_like(d_logit_real)))
+        D_loss_fake = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(logits=d_logit_fake,
+                                                                             labels=tf.zeros_like(d_logit_fake)))
+        self._D_loss = D_loss_real + D_loss_fake
+        self._G_loss = tf.reduce_mean(tf.nn.sigmoid_cross_entropy_with_logits(logits=d_logit_fake,
```
Comment (mwbrulhardt) [inline tensortrade/stochastic/processes/gan.py]: Also we should allow for loss to be defined by the user if they so choose. For example, the Wasserstein metric leads to better convergence in GANs and would be a good choice for stabilizing the training process so I should be able to give that as a loss.
